File: código/src/Partido.java
package src;

import java.util.ArrayList;

public class Partido implements Comparable<Partido>{
    private int numeroPartido; //Número do partido;
    private int votosLegenda; //Total de votos de legenda do partido;
    private String nomePartido; //Nome do partido;
    private String siglaPartido; //Sigla do partido;
    private ArrayList<Candidato> candidatosPartido; //Lista de candidatos pertencentes ao partido

    Partido(){}

    Partido(int numeroPartido, int votosLegenda, String nomePartido,String siglaPartido){
        this.numeroPartido = numeroPartido;
        this.votosLegenda = votosLegenda;
        this.nomePartido = nomePartido;
        this.siglaPartido = siglaPartido;
        this.candidatosPartido = new ArrayList<Candidato>();
    }

    //GETs e SETs

    public int getNumeroPartido(){
        return this.numeroPartido;
    }

    public void setNumeroPartido(int numeroPartido){
        this.numeroPartido = numeroPartido;
    }

    public int getVotosLegendas(){
        return this.votosLegenda;
    }

    public void setVotosLegendas(int votosLegenda){
        this.votosLegenda = votosLegenda;
    }

    public String getNomePartido(){
        return this.nomePartido;
    }

    public void setNomePartido(String nomePartido){
        this.nomePartido = nomePartido;
    }

    public String getSiglaPartido(){
        return this.siglaPartido;
    }

    public void setSiglaPartido(String siglaPartido){
        this.siglaPartido = siglaPartido;
    }  

    public ArrayList<Candidato> getCandidatoPartido(){
        return candidatosPartido;
    }

    public void setNovoCandidatoPartido(Candidato candidato){
        this.candidatosPartido.add(candidato);
    }

    public Candidato retornaCandidatoMaisVotadoPartido(){
        int maior = 0;
        Candidato candidato = new Candidato();
        for(int i=0; i < this.candidatosPartido.size();i++){
            if(this.candidatosPartido.get(i).getVotosNominais() > maior && this.candidatosPartido.get(i).getDestinoVoto().equals("Válido")){
                maior = this.candidatosPartido.get(i).getVotosNominais();
                candidato = this.candidatosPartido.get(i);
            }else{
                continue;
            }
        }
        return candidato;
    }

    public Candidato retornaCandidatoMenosVotadoPartido(String dataEleicao){
        int menor = 0, aux=0;

        Candidato candidato = new Candidato();

        for(int i=0; i < this.candidatosPartido.size();i++){
                
                if(aux==0){ //define o primeiro valor para a variavel "menor"
                    if(this.candidatosPartido.get(i).getDestinoVoto().equals("Válido")){
                        menor = this.candidatosPartido.get(i).getVotosNominais();
                        candidato = this.candidatosPartido.get(i);
                        aux++;
                        continue;
                    }
                }
                if(this.candidatosPartido.get(i).getDestinoVoto().equals("Válido")){
                    if(this.candidatosPartido.get(i).getVotosNominais() < menor){
                        menor = this.candidatosPartido.get(i).getVotosNominais();
                        candidato = this.candidatosPartido.get(i);
                    }else if(this.candidatosPartido.get(i).getVotosNominais() == menor){
                        if(this.candidatosPartido.get(i).retornaIdadeCandidato(dataEleicao) < candidato.retornaIdadeCandidato(dataEleicao)){
                            candidato = this.candidatosPartido.get(i);
                        }else if(this.candidatosPartido.get(i).retornaIdadeCandidato(dataEleicao) == candidato.retornaIdadeCandidato(dataEleicao)){
                            if(this.Candidato1MaisVelho(this.candidatosPartido.get(i), candidato) == 1){
                                candidato = this.candidatosPartido.get(i);
                            }
                        }
                    }
            }
        }
        return candidato;
    }

    public int Candidato1MaisVelho(Candidato candidato1, Candidato candidato2){
        //31/05/2001
        if(Integer.parseInt(candidato1.getDataNasc().substring(6,9)) > Integer.parseInt(candidato2.getDataNasc().substring(6,9))){
            return 0;
        }else if(Integer.parseInt(candidato1.getDataNasc().substring(6,9)) == Integer.parseInt(candidato2.getDataNasc().substring(6,9))){
            if(Integer.parseInt(candidato1.getDataNasc().substring(3,4)) > Integer.parseInt(candidato2.getDataNasc().substring(3,4))){
                return 0;
            }else if(Integer.parseInt(candidato1.getDataNasc().substring(3,4)) == Integer.parseInt(candidato2.getDataNasc().substring(3,4))){
                if(Integer.parseInt(candidato1.getDataNasc().substring(0,1)) > Integer.parseInt(candidato2.getDataNasc().substring(0,1))){
                    return 0;
                }
            }
        }
        return 1;
    }
    public int getVotosTotais(){
        int votosNominaisPartido = 0, votosTotais = 0;
        for(int i=0; i < this.candidatosPartido.size(); i++){
            if(this.candidatosPartido.get(i).getDestinoVoto().equals("Válido")){
                votosNominaisPartido += this.candidatosPartido.get(i).getVotosNominais();
            }
        }
        votosTotais = votosNominaisPartido + this.getVotosLegendas();
        return votosTotais;
    }

    @Override
    public int compareTo(Partido partido){
        // Ordenando pelos votos totais
        int votosTotaiso1 = this.getVotosTotais();
        int votosTotaiso2 = partido.getVotosTotais();

        if(votosTotaiso1 < votosTotaiso2){
            return 1;
        }else{
            return -1;
        }
    }

    @Override
    public String toString(){
        if(this.getVotosTotais() >= 2){
            if(this.getVotosTotais() - this.getVotosLegendas() >= 2){
                return String.format("%s - %d, %d votos (%d nominais e %d de legenda), ", this.siglaPartido, this.numeroPartido, this.getVotosTotais(), (this.getVotosTotais() - this.getVotosLegendas()), this.getVotosLegendas());
            }else{
                return String.format("%s - %d, %d votos (%d nominal e %d de legenda), ", this.siglaPartido, this.numeroPartido, this.getVotosTotais(), (this.getVotosTotais() - this.getVotosLegendas()), this.getVotosLegendas());
            }
        }else{
            return String.format("%s - %d, %d voto (%d nominal e %d de legenda), ", this.siglaPartido, this.numeroPartido, this.getVotosTotais(), (this.getVotosTotais() - this.getVotosLegendas()), this.getVotosLegendas());
        }
    }
}

File: TC++/código/Partido.cpp
#include "Partido.h"
#include<cstdio>
#include<cstring>

// Construtores

Partido::Partido() {};

Partido::Partido(int numeroPartido, int votosLegenda, const string& nomePartido,const string& siglaPartido,const vector<Candidato>& candidadosPartido) {
    this->numeroPartido = numeroPartido;
    this->votosLegenda = votosLegenda;
    this->nomePartido = nomePartido;
    this->siglaPartido = siglaPartido;
    this->candidatosPartido = candidatosPartido;
    this->votosNomimaisTotaisPartido = 0;
};

// GETs & SETs

void Partido::setNumeroPartido(const int numero){
    this->numeroPartido = numero;
}
const int Partido::getNumeroPartido() const{
    return this->numeroPartido;
}

void Partido::setVotosLegenda(int votosLegenda){
    this->votosLegenda = votosLegenda;
}
const int Partido::getVotosLegenda() const{
    return this->votosLegenda;
}   

void Partido::setVotosNominaisTotal(int votos){
    this->votosNomimaisTotaisPartido += votos;
}
const int Partido::getVotosNominaisTotal() const{
    return this->votosNomimaisTotaisPartido;
}

void Partido::setNomePartido(const string &nome){
    this->nomePartido = nome;
}
const string& Partido::getNomePartido() const{
    return this->nomePartido;
}   

void Partido::setSiglaPartido(const string &sigla){
    this->siglaPartido = sigla;
}
const string& Partido::getSiglaPartido() const{
    return this->siglaPartido;
}   

void Partido::setNovoCandidatoPartido(const Candidato &candidato){
    this->candidatosPartido.push_back(candidato);
}
const vector <Candidato>& Partido::getCandidatoPartido(){
    return this->candidatosPartido;
}

void Partido::setPos(int pos){
    this->pos = pos;
}
const int Partido::getPos() const{
    return this->pos;
}

// Métodos

const int Partido::getVotosTotais() const{
    int votosNominaisPartido = 0, votosTotais = 0;
    for(std::vector<Candidato>::size_type i=0; i < this->candidatosPartido.size(); i++){
        if(this->candidatosPartido[i].getDestinoVoto() == "Válido"){
            votosNominaisPartido += this->candidatosPartido[i].getVotosNominais();
        }
    }
    votosTotais = votosNominaisPartido + this->getVotosLegenda();
    return votosTotais;
}

const Candidato Partido::retornaCandidatoMaisVotadoPartido() const{
    int maior = 0;
    Candidato candidato;
    for(std::vector<Candidato>::size_type i=0; i < this->candidatosPartido.size();i++){
        if(this->candidatosPartido[i].getVotosNominais() > maior &&  this->candidatosPartido[i].getDestinoVoto() == "Válido"){
            maior = this->candidatosPartido[i].getVotosNominais();
            candidato = this->candidatosPartido[i];
        }else{
            continue;
        }
    }
    return candidato;
}

const Candidato Partido::retornaCandidatoMenosVotadoPartido(const string& dataEleicao) const{
    int menor = 0, aux=0;

    Candidato candidato;

    for(std::vector<Candidato>::size_type i=0; i < this->candidatosPartido.size();i++){
            
        if(aux==0){ //define o primeiro valor para a variavel "menor"
            if(this->candidatosPartido[i].getDestinoVoto() == "Válido" ){
                menor = this->candidatosPartido[i].getVotosNominais();
                candidato = this->candidatosPartido[i];
                aux++;
                continue;
            }
        }
        if(this->candidatosPartido[i].getDestinoVoto() == "Válido"){
            if(this->candidatosPartido[i].getVotosNominais() < menor){
                menor = this->candidatosPartido[i].getVotosNominais();
                candidato = this->candidatosPartido[i];
            }
            else if(this->candidatosPartido[i].getVotosNominais() == menor){
                if(this->candidatosPartido[i].retornaIdadeCandidato(dataEleicao) < candidato.retornaIdadeCandidato(dataEleicao)){
                    candidato = this->candidatosPartido[i];
                }
            }else if(this->candidatosPartido[i].retornaIdadeCandidato(dataEleicao) == candidato.retornaIdadeCandidato(dataEleicao)){
                if(this->Candidato1MaisVelho(this->candidatosPartido[i], candidato) == 1){
                    candidato = this->candidatosPartido[i];
                }
            }
        }
    }
    return candidato;
}

const int Partido::Candidato1MaisVelho(const Candidato& candidato1,const Candidato& candidato2) const{
    //31/05/2001
    if(stoi(candidato1.getDataNasc().substr(6,9)) > stoi(candidato2.getDataNasc().substr(6,9))){
        return 0;
    }else if(stoi(candidato1.getDataNasc().substr(6,9)) == stoi(candidato2.getDataNasc().substr(6,9))){
        if(stoi(candidato1.getDataNasc().substr(3,4)) > stoi(candidato2.getDataNasc().substr(3,4))){
            return 0;
        }else if(stoi(candidato1.getDataNasc().substr(3,4)) == stoi(candidato2.getDataNasc().substr(3,4))){
            if(stoi(candidato1.getDataNasc().substr(0,1)) > stoi(candidato2.getDataNasc().substr(0,1))){
                return 0;
            }
        }
    }
    return 1;
}

ostream& operator<< (ostream &out, const Partido &p1){
    if(p1.getVotosTotais() >= 2){
        if(p1.getVotosTotais() - p1.getVotosLegenda() >= 2){
            return out << p1.siglaPartido << " - " << p1.numeroPartido << ", " << p1.getVotosTotais() << " votos " << "(" << (p1.getVotosTotais() - p1.getVotosLegenda()) << " nominais e " << p1.getVotosLegenda() << " de legenda)" << ", ";
        }else{
            return out << p1.siglaPartido << " - " << p1.numeroPartido << ", " << p1.getVotosTotais() << " votos " << "(" << (p1.getVotosTotais() - p1.getVotosLegenda()) << " nominal e " << p1.getVotosLegenda() << " de legenda)" << ", ";
        }
    }else{
        return out << p1.siglaPartido << " - " << p1.numeroPartido << ", " << p1.getVotosTotais() << " voto " << "(" << (p1.getVotosTotais() - p1.getVotosLegenda()) << " nominal e " << p1.getVotosLegenda() << " de legenda)" << ", ";
    }
}

bool Partido::compareVotosTotais(const Partido& p1, const Partido& p2){
    if(p1.getVotosTotais() > p2.getVotosTotais()){
        return true;
    }else if(p1.getVotosTotais() < p2.getVotosTotais()){
        return false;
    }else{
        return 0;
    }
}